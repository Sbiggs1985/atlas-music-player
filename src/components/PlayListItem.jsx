import React from "react";

export default function PlayListItem({ title, artist, duration, isSelected, onClick }) {
  const formatTime = (seconds) => {
    const minutes = Math.floor(seconds / 60);
    const remainingSeconds = Math.floor(seconds % 60);
    return `${minutes}:${remainingSeconds < 10 ? "0" : ""}${remainingSeconds}`;
  };

  return (
    <div
      className={`flex justify-between items-center p-4 rounded-lg cursor-pointer ${
        isSelected ? "bg-blue-500 text-white" : "hover:bg-gray-100"
      }`}
      onClick={onClick}
    >
      {/* Song Title and Artist */}
      <div>
        <p className="text-lg font-bold">{title}</p>
        <p className="text-sm text-gray-500">{artist}</p>
      </div>

      {/* Duration */}
      <div className="text-sm text-gray-500">
        {formatTime(duration)}
      </div>
    </div>
  );
}