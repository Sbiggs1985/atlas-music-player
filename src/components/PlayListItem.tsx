import React from "react";

// Define the props type
interface PlayListItemProps {
  title: string; // Song title
  artist: string; // Artist name
  duration: number; // Duration in seconds
  isSelected: boolean; // Indicates if the item is currently selected
  onClick: () => void; // Click handler
}

const PlayListItem: React.FC<PlayListItemProps> = ({
  title,
  artist,
  duration,
  isSelected,
  onClick,
}) => {
  // Format duration from seconds to HH:MM format
  const formatTime = (seconds: number | null | undefined): string => {
    if (!seconds) return "0:00"; // Handle cases where duration is missing

    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const remainingSeconds = Math.floor(seconds % 60);

    return hours > 0
      ? `${hours}:${minutes.toString().padStart(2, "0")}:${remainingSeconds
          .toString()
          .padStart(2, "0")}`
      : `${minutes}:${remainingSeconds.toString().padStart(2, "0")}`;
  };

  return (
    <div
      className={`flex justify-between items-center p-4 rounded-lg cursor-pointer 
        ${
          isSelected
            ? "bg-blue-500 text-white dark:bg-blue-600"
            : "hover:bg-gray-200 dark:hover:bg-gray-700"
        }`}
      onClick={onClick}
      role="button"
      aria-pressed={isSelected}
      aria-selected={isSelected}
      tabIndex={0}
    >
      {/* Song Title and Artist */}
      <div>
        <p className="text-lg font-bold">{title}</p>
        <p className="text-sm text-gray-600 dark:text-gray-400">{artist}</p>
      </div>

      {/* Duration */}
      <div className="text-sm text-gray-600 dark:text-gray-400">
        {formatTime(duration)}
      </div>
    </div>
  );
};

export default PlayListItem;

// Props type: Added a PlayListItemProps interface to define the types for props:
//  * title(string), artist(string), duration(number), isSelected(boolean).