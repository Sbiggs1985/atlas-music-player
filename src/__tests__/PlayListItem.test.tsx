import { render, screen, fireEvent } from "@testing-library/react";
import PlayListItem from "@/components/PlayListItem";
import { describe, it, expect, vi } from "vitest";

describe("PlayListItem Component", () => {
  it("renders the song title and artist correctly", () => {
    render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={false} onClick={() => {}} />);

    expect(screen.getByText("Test Song")).toBeInTheDocument();
    expect(screen.getByText("Test Artist")).toBeInTheDocument();
  });

  it("formats and displays the duration correctly", () => {
    render(<PlayListItem title="Test Song" artist="Test Artist" duration={3945} isSelected={false} onClick={() => {}} />);
    
    // Expect 1 hour, 5 minutes, and 45 seconds
    expect(screen.getByText("1:05:45")).toBeInTheDocument();
  });

  it("applies correct styles when selected", () => {
    render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={true} onClick={() => {}} />);
    
    const listItem = screen.getByRole("button");
    expect(listItem).toHaveClass("bg-blue-500 text-white");
  });

  it("applies hover styles when not selected", () => {
    render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={false} onClick={() => {}} />);
    
    const listItem = screen.getByRole("button");
    expect(listItem).toHaveClass("hover:bg-gray-100");
  });

  it("calls the onClick function when clicked", () => {
    const onClickMock = vi.fn();
    render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={false} onClick={onClickMock} />);

    const listItem = screen.getByRole("button");
    fireEvent.click(listItem);
    
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot when selected", () => {
    const { asFragment } = render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={true} onClick={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when not selected", () => {
    const { asFragment } = render(<PlayListItem title="Test Song" artist="Test Artist" duration={210} isSelected={false} onClick={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});