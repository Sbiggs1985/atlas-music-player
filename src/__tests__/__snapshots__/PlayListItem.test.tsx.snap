// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayListItem Component > matches snapshot when not selected 1`] = `
<DocumentFragment>
  <div
    aria-pressed="false"
    class="flex justify-between items-center p-4 rounded-lg cursor-pointer hover:bg-gray-100"
    role="button"
    tabindex="0"
  >
    <div>
      <p
        class="text-lg font-bold"
      >
        Test Song
      </p>
      <p
        class="text-sm text-gray-500"
      >
        Test Artist
      </p>
    </div>
    <div
      class="text-sm text-gray-500"
    >
      3:30
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlayListItem Component > matches snapshot when selected 1`] = `
<DocumentFragment>
  <div
    aria-pressed="true"
    class="flex justify-between items-center p-4 rounded-lg cursor-pointer bg-blue-500 text-white"
    role="button"
    tabindex="0"
  >
    <div>
      <p
        class="text-lg font-bold"
      >
        Test Song
      </p>
      <p
        class="text-sm text-gray-500"
      >
        Test Artist
      </p>
    </div>
    <div
      class="text-sm text-gray-500"
    >
      3:30
    </div>
  </div>
</DocumentFragment>
`;
